IdentityCore/src/webview/systemWebview/ios/MSIDSystemWebviewController.m

```diff
@@ -64,34 +69,59 @@ - (instancetype)initWithStartURL:(NSURL *)startURL
         _startURL = startURL;
         _context = context;
         _callbackURLScheme = callbackURLScheme;
+        _parentController = parentController;
+        _allowSafariViewController = allowSafariViewController;
+        _useAuthenticationSession = useAuthenticationSession;
     }
-    
     return self;
 }
 
 - (void)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
 {
-    if (@available(iOS 11.0, *))
+    if (!completionHandler)
     {
-        _session = [[MSIDSFAuthenticationSession alloc] initWithURL:self.startURL
-                                                  callbackURLScheme:self.callbackURLScheme
-                                                            context:_context];
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"CompletionHandler cannot be nil for interactive session.", nil, nil, nil, _context.correlationId, nil);
+        [MSIDNotifications notifyWebAuthDidFailWithError:error];
+        completionHandler(nil, error);
+        return;
+    }
 
+    NSError *error = nil;
+    
+    if (_useAuthenticationSession)
+    {
+        if (@available(iOS 11.0, *))
+        {
+            _session = [[MSIDAuthenticationSession alloc] initWithURL:self.startURL
+                                                    callbackURLScheme:self.callbackURLScheme
+                                                              context:_context];
+        }
+        else
+        {
+             error = MSIDCreateError(MSIDErrorDomain, MSIDErrorUnsupportedFunctionality, @"SFAuthenticationSession/ASWebAuthenticationSession is not available for iOS 10 and older.", nil, nil, nil, _context.correlationId, nil);
+        }
     }
-    else
+    
+    if (!_session && _allowSafariViewController)
     {
         _session = [[MSIDSafariViewController alloc] initWithURL:_startURL
+                                                parentController:_parentController
                                                          context:_context];
     }
     
-    if (_session)
+    if (!_session)
     {
-        [_session startWithCompletionHandler:completionHandler];
+        if (!error)
+        {
+            error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractiveSessionStartFailure, @"Failed to create an auth session", nil, nil, nil, _context.correlationId, nil);
+        }
+        [MSIDNotifications notifyWebAuthDidFailWithError:error];
+        completionHandler(nil, error);
         return;
     }
     
-    NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractiveSessionStartFailure, @"Failed to create an auth session", nil, nil, nil, _context.correlationId, nil);
-    completionHandler(nil, error);
+    [MSIDNotifications notifyWebAuthDidStartLoad:_startURL];
+    [_session startWithCompletionHandler:completionHandler];
```

Comment:
what if session fails to start?

Reply:
then corresponding completionHandler would be called, unless completionHandler is nil. But such case is already detected beforehand in MSIDWebviewAuthorization.